lib/core/mqtt.cpp

```diff
@@ -0,0 +1,174 @@
+#include "mqtt.hpp"
+
+#include <rapidjson/writer.h>
+
+#include "rapidjson/stringbuffer.h"
+
+namespace hyped::core {
+
+std::optional<std::shared_ptr<Mqtt>> Mqtt::create(ILogger &logger,
+                                                  const std::string &id,
+                                                  const std::string &host,
+                                                  const std::uint16_t port)
+{
+  std::string address = "tcp://" + host + ":" + std::to_string(port);
+  mqtt::connect_options connection_options;
+  connection_options.set_keep_alive_interval(1);
+  connection_options.set_clean_session(true);
+  auto mqtt_client = std::make_unique<mqtt::client>(address, id);
+  auto cb          = std::make_shared<MqttCallback>(logger);
+  mqtt_client->set_callback(*cb);
+  mqtt_client->connect(connection_options);
+  if (!mqtt_client->is_connected()) { return std::nullopt; }
+  logger.log(core::LogLevel::kInfo, "Successfully created MQTT client and connected to broker");
+  return std::make_shared<Mqtt>(logger, std::move(mqtt_client), std::move(cb));
+}
+
+Mqtt::Mqtt(ILogger &logger, std::unique_ptr<mqtt::client> client, mqtt::callback_ptr cb)
+    : logger_(logger),
+      client_(std::move(client)),
+      messages_in_queue(0),
+      cb(std::move(cb))
+{
+}
+
+Mqtt::~Mqtt()
+{
+  client_->disconnect();
+}
+
+void Mqtt::publish(const MqttMessage &message, const MqttMessageQos qos)
+{
+  auto constructed_message = messageToMessagePtr(message);
+  logger_.log(core::LogLevel::kDebug,
+              "Publishing message %s",
+              constructed_message->get_payload_str().c_str());
+  constructed_message->set_qos(qos);
+  client_->publish(constructed_message);
+}
+
+core::Result Mqtt::subscribe(const core::MqttTopic topic)
+{
+  const auto topic_string = mqtt_topic_to_string.find(topic);
+  if (topic_string == mqtt_topic_to_string.end()) {
+    logger_.log(core::LogLevel::kFatal, "Attempted to subscribe to nonexistent MQTT topic");
+    return core::Result::kSuccess;
```

Comment:
also, for a conversation topic, i feel like this is the kind of thing we should catch at compile time with `static_assert`, as I believe we would never have a run time configurable topic? Would like opinions

Reply:
would that work as a `static_assert` inside subscribe? That feels more elegant

Reply:
acc on second thoughts is it not impossible for us to subscribe to an undefined topic? As the topic passed in is a member of `core::MqttTopic`, so a topic has to be defined for us to even pass it into the function. I may be thinking about this wrong though

Reply:
can we `static_assert` that the topic is in `mqtt_topic_to_string`? if we can it would be nice to get a compile time error that we had forgotten to add it to the hashmap

Reply:
cool will add that